src/data/fin.lean

```diff
@@ -991,6 +1005,57 @@ by { ext j, simp [init] }
 
 end tuple_right
 
+section insert_nth
+
+variables {α : fin (n+1) → Type u} {β : Type v}
+
+/-- Insert an element into a tuple at a given position. For `i = 0` see `fin.cons`,
+for `i = fin.last n` see `fin.snoc`. -/
+def insert_nth (i : fin (n + 1)) (x : α i) (p : Π j : fin n, α (i.succ_above j))
+  (j : fin (n + 1)) : α j :=
+if h : j = i
+then _root_.cast (congr_arg α h.symm) x
+else _root_.cast (congr_arg α $ succ_above_pred_above i _ h) (p $ i.pred_above j h)
+
+@[simp] lemma insert_nth_apply_same (i : fin (n + 1)) (x : α i) (p : Π j, α (i.succ_above j)) :
+  insert_nth i x p i = x :=
+by simp [insert_nth]
+
+@[simp] lemma insert_nth_apply_succ_above (i : fin (n + 1)) (x : α i) (p : Π j, α (i.succ_above j))
+  (j : fin n) :
+  insert_nth i x p (i.succ_above j) = p j :=
+begin
+  simp only [insert_nth, dif_neg (succ_above_ne _ _)],
+  refine eq_of_heq ((cast_heq _ _).trans _),
+  rw [pred_above_succ_above]
+end
+
+@[simp] lemma insert_nth_comp_succ_above (i : fin (n + 1)) (x : β) (p : fin n → β) :
+  insert_nth i x p ∘ i.succ_above = p :=
+funext $ insert_nth_apply_succ_above i x p
+
+lemma insert_nth_eq_iff {i : fin (n + 1)} {x : α i} {p : Π j, α (i.succ_above j)} {q : Π j, α j} :
+  i.insert_nth x p = q ↔ q i = x ∧ p = (λ j, q (i.succ_above j)) :=
+by simp [funext_iff, forall_iff_succ_above i, eq_comm]
+
+lemma eq_insert_nth_iff {i : fin (n + 1)} {x : α i} {p : Π j, α (i.succ_above j)} {q : Π j, α j} :
+  q = i.insert_nth x p ↔ q i = x ∧ p = (λ j, q (i.succ_above j)) :=
+eq_comm.trans insert_nth_eq_iff
+
+lemma insert_nth_zero (x : α 0) (p : Π j : fin n, α (succ_above 0 j)) :
+  insert_nth 0 x p = cons x (λ j, _root_.cast (congr_arg α (congr_fun succ_above_zero j)) (p j)) :=
+begin
+  refine insert_nth_eq_iff.2 ⟨by simp, _⟩,
+  ext j,
+  convert (cons_succ _ _ _).symm
+end
+
+@[simp] lemma insert_nth_zero' {α : Type*} (x : α) (p : fin n → α) :
+  @insert_nth _ (λ _, α) 0 x p = cons x p :=
+by simp [insert_nth_zero]
```

Comment:
Can you define a similar lemma for `snoc`?  At a guess
``
@[simp] lemma insert_nth_n {α : Type*} (x : α) (p : fin n → α) :
  @insert_nth _ (λ _, α) n x p = snoc x p :=
```

Reply:
Done.